index.js

```diff
@@ -1,8 +1,60 @@
-var bus = require('statebus').serve({file_store: false})
-bus.honk = false;
-var serve = (route, file) => bus.http.get(route, (r, res) => res.sendFile(__dirname + file))
-
-serve('/janus.js', '/janus.js')
-serve('/hark.js',  '/node_modules/hark/hark.bundle.js')
-serve('/',         '/index.html')
-serve('/:id',      '/index.html')
+'use strict';
+
+function get_spaces(conns) {
+  // connections state on the server is different than the client.
+  // It looks like {key: 'connections', '<random-hash>': {...}, '<another-hash>': {...}}
+  const spaces = new Set();
+  for (let connid in conns) {
+    if (connid != 'key' && conns[connid].space !== undefined) {
+      spaces.add(conns[connid].space);
+    }
+  }
+  return spaces;
+}
+
+function set_difference(a, b) {
+  return new Set([...a].filter(x => !b.has(x)));
+}
+
+function add_server_state(bus) {
+  let previous_spaces = new Set();
+  // If the db gets lost, restart this counter at 1000
+  bus.state.chats_served = bus.state.chats_served || 1000;
+
+  // TODO(karth295): This should be more efficient than
+  // computing all spaces every time. Ideally when every
+  // connection is added or removed we just update the data structures
+  // to reflect the change from that one connection.
+  bus.reactive(()=> {
+    const new_spaces = get_spaces(bus.fetch('connections'));
+
+    const diff = set_difference(previous_spaces, new_spaces);
+    bus.state.chats_served += diff.size;
+
+    previous_spaces = new_spaces;
+  })();
```

Comment:
Hey I should have mentioned this before. The purpose of this reactive function is to *define* the state `chats_served`, and you can define that directly with a `.to_fetch` function:

```javascript
bus('chats_served').to_fetch = (old) => {
  const new_spaces = get_spaces(bus.fetch('connections'))
  const diff = set_difference(previous_spaces, new_spaces)
  previous_spaces = new_spaces
  return {_: old._ + diff}
}
```

I think we need a better name than .to_fetch... perhaps .define.

Reply:
@toomim thanks -- I didn't realize to_fetch gave you the old value -- I thought it was the key that gets fetched.

Question: is there a way to hook into the `to_save` function of `/connection` from a statebus server?. I realize this code would be far more efficient if we avoided recomputing these sets every single time a connection is saved. Aka every time somebody moves their cursor.

Reply:
@karth295, the to_fetch function looks at its parameter name to figure out what to give it. You can do any of these:

```javascript
bus('chats_served').to_fetch = (old) => {
```
```javascript
bus('chats_served').to_fetch = (key) => {
```
```javascript
bus('chats_served').to_fetch = (k, old) => {
```
```javascript
bus('chats_served').to_fetch = (key, old) => {
```
```javascript
bus('chats_served').to_fetch = (old, k) => {
```

Reply:
@karth295 you can define a `master('connections').to_save = (obj) => {...}` function, sure. How are you thinking of making it more efficient?

Reply:
Right now, every time a connection is updated in any form, we need to do an O(connections) operation to determine whether to increment this `chats_served` counter. I think it could be O(1) if there were a way to get a callback for a specific connection changing. Like `master('connection/*').to_save`).

Alternatively, is this possible through the client-specific bus?

Reply:
> Also, we'll eventually be adding handlers for specific types of diffs, so that you can get a callback just when the list of connections gets smaller or larger.

Actually, we've already got enough [diff support in statebus](https://wiki.invisible.college/statebus/diffs) that we could implement this now. We'd modify the implementation of the `connections` state in `server.js` so that each time it adds or removes a connection, it also communicates the patch. For instance, we'd change this code in server.js:

```javascript
                // Close connection
                if (user_bus_func) {
                    delete connections[conn.id]; master.save(connections)
                    user.delete_bus()
                }
```

to this:

```javascript
                // Close connection
                if (user_bus_func) {
                    delete connections[conn.id]
                    master.save('connections', {patch: `delete .${conn.id}`})
                    user.delete_bus()
                }
```

And we'd replace this:

```javascript
                // Open connection
                connections[conn.id] = {client: conn.id}; master.save(connections)
```

with this:
```javascript
                // Open connection
                connections[conn.id] = {client: conn.id}
                master.save('connections', {patch: `.${conn.id} = {"client": "${conn.id}"}`})
```

Then we can look for the patch in a `master('connections').on_save = (o, t) => { if (t.patch) ... }` handler.

Now, this probably isn't worth doing right now. But I just wanted to take the opportunity to explore some of the future power we're developing with the diff/sync/patch architecture.

Reply:
Oh, and augmenting saves with these patches might not be worth doing *ever* — because they will be auto-generated in the future, once we get the new JSON encoding, and revamp the `state` proxy object to use it, we can have the `state` object auto-generate a patch for each change you make.

So we can just wait until that's done, and then we'll be able to get O(1) updates without much additional work.